rfd/0107-desktop-access-performance-research.md

```diff
@@ -0,0 +1,95 @@
+The flow of the messages is controlled by the server so it is not possible to have a message which renders a bitmap that uses cache key that hasn't been yet used to store a bitmap. 
+
+#### What changes are required to support caching?
+
+Teleport Desktop Access protocol, TDP, needs to be extended with two messages: 
+
+#### 29 - bitmap cache store
+
+This message is sent from the server to the client to store a bitmap in the bitmap cache
+
+```
+| message type (29) | cache_id uint32 | cache_index uint32 | data_length uint32 | data []byte |
+```
+
+#### 30 - bitmap cache load
+
+This message is sent from the server to the client to render a bitmap stored in the bitmap cache
+
+```
+| message type (30) | cache_id uint32 | cache_index uint32 | top uint32 | left uint32 | right uint32 | bottom uint32 |
+```
+
+Update to the UI is also required. The browser memory and javascript's array of hash tables, where cache id is the index of the array, and the cache index is a key in the hash table, will be used to store bitmap data.
+
+
+There's a sequence diagram which shows the flow of the messages between all components:
+
+```mermaid
+sequenceDiagram
+    Windows Server->>Teleport Windows Desktop Service: rdp: store bitmap
+    Teleport Windows Desktop Service->>Teleport Proxy: tdp: store bitmap
+    Teleport Proxy->>Web Browser: tdp: store bitmap
+    Web Browser->>cache in web browser memory: store bitmap
+    Windows Server->>Teleport Windows Desktop Service: rdp: load bitmap
+    Teleport Windows Desktop Service->>Teleport Proxy: tdp: load bitmap
+    Teleport Proxy->>Web Browser: tdp: load bitmap
+    cache in web browser memory->>Web Browser: load bitmap
+```
+
+- server sends [message](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/3c76f8fb-fcaa-4ca6-9d2e-3f449bdbed40) to inform client to store bitmap in the cache
+- windows destkop service translates rdp message into tdp message to store bitmap and sends it to the proxy
+- proxy forwards message to web browser
+- web browser stores bitmap in the in-memory cache
+- server sends [message](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/84c2ec2f-f776-405b-9b48-6894a28b1b14) to inform client to load bitmap from the cache
+- windows destkop service translates rdp message into tdp message to load bitmap and sends it to the proxy
+- proxy forwards message to web browser
+- web browser loads bitmap from the in-memory cache
+
+#### session recordings
+Since we no longer send every bitmap over the wire and to keep session recordings to work we'll also need to keep the cache of the bitmaps at the proxy. When we encounter the `save bitmap message` we'll need to store that bitmap at the proxy and then whenver `load bitmap message` will be sent we'll need to load bitmap from the cache and generate appropriate event with the bitmap data. 
```

Comment:
Is this necessary?

If we record the messages that save to and load from the cache, we can play back the recording in the same way, right?

Reply:
It's really not necessary. The other option works as well.When I was thinking about what needs to be changed to make session recording work with new cache messages it looked like it might be an easier thing to do. The more I think about it now, the option with updating the logic for session recording will be cleaner solution, as we won't need to store anything on the proxy level. We'll  only need to update the function for recordings. Anyway, thanks for suggestion! I'll update this section.